python/ray/serve/master.py

```diff
@@ -29,6 +29,42 @@
 _RESOURCE_CHECK_ENABLED = True
 
 
+class TrafficPolicy:
+    def __init__(self, traffic_dict):
+        self.traffic_dict = dict()
+        self.shadow_dict = dict()
+        self.set_traffic_dict(traffic_dict)
+
+    def set_traffic_dict(self, traffic_dict):
+        prob = 0
+        for backend, weight in traffic_dict.items():
+            if weight < 0:
+                raise ValueError(
+                    "Attempted to assign a weight of {} to backend '{}'. "
+                    "Weights cannot be negative.".format(weight, backend))
+            prob += weight
+
+        # These weights will later be plugged into np.random.choice, which
+        # uses a tolerance of 1e-8.
+        if not np.isclose(prob, 1, atol=1e-8):
+            raise ValueError("Traffic dictionary weights must sum to 1, "
+                             "currently they sum to {}".format(prob))
+        self.traffic_dict = traffic_dict
+
+    def set_shadow(self, backend, proportion):
+        if proportion == 0 and backend in self.shadow_dict:
+            del self.shadow_dict[backend]
+        else:
+            self.shadow_dict[backend] = proportion
+
+
+class BackendInfo:
```

Comment:
maybe use namedtuple instead of this kind of shallow class?

Reply:
Oh nice, I didn't know this exists